File: server/internal/content/ip-profile/scope.go
package ipprofile

import (
	"errors"
	"maps"
	"slices"
)

// The account's material scope preference: which sources a run starts from.
//
// This is "what was chosen last time", not a configuration version. The persona
// prompt needs a version history because a run pins one (LT-012); this needs
// none, because every run already records the preference it saw in its own
// snapshot (diagnostics.Snapshot.Preference).
//
// It lives at settings["loretide.scope"] rather than in a column, so there is
// no migration and no change to the account table. The shape is taken from the
// brand workspace's timezone (specs/004): validate on write, fill the default
// on read, and never let a read write.
//
// Contract: specs/018-lt014-account-scope-preference/contracts/account-scope.md

// Scope is where a run may draw material from.
type Scope string

const (
	// ScopeLocal uses only material already held locally.
	ScopeLocal Scope = "local"
	// ScopeWeb uses only material retrieved from the web.
	ScopeWeb Scope = "web"
	// ScopeAll uses both.
	ScopeAll Scope = "all"
)

// Scopes is the controlled set. Mirrored by packages/core/content/ip-profile/
// scope.ts, which reads this file and compares rather than restating it.
var Scopes = []Scope{ScopeLocal, ScopeWeb, ScopeAll}

// ScopeSettingsKey is where the preference lives inside the account's settings
// JSONB. The "loretide." prefix keeps it clear of any upstream key called
// "scope", the same reason the workspace timezone carries one.
const ScopeSettingsKey = "loretide.scope"

// DefaultScope is what an account reads as when it has never chosen - including
// every account created before this feature. Never the empty string: the start
// screen renders a radio group from this value, and an empty one would leave
// all three unselected on first open.
const DefaultScope = string(ScopeAll)

// ErrScope is an unrecognised scope. Answered as 400 with a diagnostic error
// object, never as a storage error.
var ErrScope = errors.New("unsupported material scope")

// ValidateScope accepts only an exact match: no trimming, no case folding.
// Tolerating "All" alongside "all" would give one preference two spellings in
// storage, and every reader would then have to know both.
func ValidateScope(value string) error {
	if !slices.Contains(Scopes, Scope(value)) {
		return ErrScope
	}
	return nil
}

// ScopeOf reads the preference off a settings blob, falling back to the default
// for anything unreadable: settings that are absent, not an object, missing the
// key, or holding a value nothing recognises.
//
// It does NOT write the default back, and callers must not either. An account
// that never chose and one that chose "all" deliberately would otherwise become
// the same row, and the first read would have made that choice for the person.
func ScopeOf(settings map[string]any) string {
	raw, present := settings[ScopeSettingsKey]
	if !present {
		return DefaultScope
	}
	value, isString := raw.(string)
	if !isString || ValidateScope(value) != nil {
		return DefaultScope
	}
	return value
}

// withScope returns a NEW settings map carrying the scope, leaving the caller's
// untouched.
//
// Merging is not an optimisation. The update query replaces the settings blob
// wholesale (handler/content_account.go marshals the patch straight over the
// stored value), so writing only this key would delete every other setting the
// account has. Doing it here means no call site has to remember that.
func withScope(settings map[string]any, scope string) map[string]any {
	merged := make(map[string]any, len(settings)+1)
	maps.Copy(merged, settings)
	merged[ScopeSettingsKey] = scope
	return merged
}

// scopeFilled returns settings with the scope key present, filling the default
// when it is absent or unusable. Applied on the way OUT only: the stored row
// keeps whatever it had, and only the response is complete.
func scopeFilled(settings map[string]any) map[string]any {
	if existing, present := settings[ScopeSettingsKey]; present {
		if value, isString := existing.(string); isString && ValidateScope(value) == nil {
			return settings
		}
	}
	return withScope(settings, DefaultScope)
}

// validateScopeSetting rejects a settings blob whose scope key is unusable.
// Settings without the key are fine: the preference is optional at creation,
// and an unrelated update must not be forced to carry one.
//
// This runs on the general update path as well as the dedicated endpoint. With
// only the endpoint checking, anyone could put a value the readers refuse into
// the row through PATCH, and the endpoint's validation would be decoration.
func validateScopeSetting(settings map[string]any) error {
	raw, present := settings[ScopeSettingsKey]
	if !present {
		return nil
	}
	value, isString := raw.(string)
	if !isString {
		return ErrScope
	}
	return ValidateScope(value)
}

File: packages/core/content/ip-profile/scope.ts
// The account's material scope preference: which sources a run starts from.
//
// The single source of truth is Go: `Scopes` in
// `server/internal/content/ip-profile/scope.go`. This file is the frontend's
// copy, and `scope.test.ts` reads that file and compares, so the two cannot
// drift apart silently.
//
// The value lives at settings["loretide.scope"] on the account, the same shape
// the brand workspace's timezone uses (packages/core/workspace/timezone.ts):
// the default is filled in on READ and never written back, so "never chose" and
// "chose all" stay different states.

export const CONTENT_SCOPES = ["local", "web", "all"] as const;

export type ContentScope = (typeof CONTENT_SCOPES)[number];

/** Where the preference lives inside the account's settings JSON. The
 *  "loretide." prefix keeps it clear of any upstream key called "scope". */
export const SCOPE_SETTINGS_KEY = "loretide.scope";

/** What an account reads as when it has never chosen — including every account
 *  created before this feature. Never an empty string: the start screen renders
 *  a radio group from this value. */
export const DEFAULT_SCOPE: ContentScope = "all";

export function isContentScope(value: unknown): value is ContentScope {
  return (
    typeof value === "string" &&
    (CONTENT_SCOPES as readonly string[]).includes(value)
  );
}

/**
 * Reads the preference off an account, falling back to the default for anything
 * unreadable: no settings, settings that are not an object, a missing key, or a
 * value this build does not recognise.
 *
 * Every fallback here is reachable from a real response — an account created
 * before this feature, a server that sent settings as null, a blob written by
 * code that did not know the key — so per constitution principle VI the caller
 * does not have to guard each one.
 */
export function getAccountScope(
  account: { settings?: unknown } | undefined,
): ContentScope {
  const settings = account?.settings;
  if (typeof settings !== "object" || settings === null || Array.isArray(settings)) {
    return DEFAULT_SCOPE;
  }
  const value = (settings as Record<string, unknown>)[SCOPE_SETTINGS_KEY];
  return isContentScope(value) ? value : DEFAULT_SCOPE;
}

/**
 * Whether the account has a stored preference of its own.
 *
 * Distinct from getAccountScope, which always returns something: an account
 * that never chose and one that deliberately chose "all" read the same there.
 */
export function hasStoredScope(account: { settings?: unknown } | undefined): boolean {
  const settings = account?.settings;
  if (typeof settings !== "object" || settings === null || Array.isArray(settings)) {
    return false;
  }
  return isContentScope((settings as Record<string, unknown>)[SCOPE_SETTINGS_KEY]);
}

/**
 * Merges a scope into an existing settings object and returns a new one.
 *
 * The dedicated endpoint merges server-side, so nothing in this package needs
 * this to write. It exists for a caller that has an account in hand and wants
 * the settings it WOULD have — and it merges rather than replaces for the same
 * reason the server does: PATCH replaces the blob wholesale, so a partial
 * object sent there would delete the account's other settings.
 *
 * Throws on an unrecognised scope rather than producing a blob no reader
 * accepts.
 */
export function withAccountScope(
  settings: unknown,
  scope: string,
): Record<string, unknown> {
  if (!isContentScope(scope)) {
    throw new Error(`invalid material scope: ${scope}`);
  }
  const base =
    typeof settings === "object" && settings !== null && !Array.isArray(settings)
      ? (settings as Record<string, unknown>)
      : {};
  return { ...base, [SCOPE_SETTINGS_KEY]: scope };
}
